cuda/matrix/dense_kernels.cu

```diff
@@ -427,28 +634,145 @@ void move_to_sellp(std::shared_ptr<const CudaExecutor> exec,
 GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_DENSE_MOVE_TO_SELLP_KERNEL);
 
+namespace kernel{
+
+__global__ __launch_bounds__(default_block_size) void reduce_total_nnz(
+    size_type num_rows, size_type* nnz_per_row, size_type* result, size_type start_step)
+{
+	const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
+	for (auto step = start_step; step >= 1; step/=2) {
+		if (tidx < step && tidx + step < num_rows) {
+			nnz_per_row[tidx] += nnz_per_row[tidx + step];
+		}
+		__syncthreads();
+	}
+	if (tidx == 0) *result = nnz_per_row[0];
+}
+
+} // namespace kernel
+
 
 template <typename ValueType>
 void count_nonzeros(std::shared_ptr<const CudaExecutor> exec,
                     const matrix::Dense<ValueType> *source,
-                    size_type *result) NOT_IMPLEMENTED;
+                    size_type *result)
+{
+	const auto num_rows = source->get_size()[0];
+	auto tmp = gko::Array<size_type>(exec);
+	tmp.resize_and_reset(num_rows);
+
+	calculate_nonzeros_per_row(exec, source, &tmp);
+
+	size_type* d_result;
+
+	cudaMalloc(&d_result, sizeof(size_type));
+	auto start_step = 1;
+	while (2 * start_step < num_rows) start_step *= 2;
+
+	auto grid_dim = ceildiv(num_rows, default_block_size);
+	kernel::reduce_total_nnz<<<grid_dim, default_block_size>>>(num_rows, as_cuda_type(tmp.get_data()), d_result,
+			start_step);
+
+	cudaMemcpy(result, d_result, sizeof(size_type), cudaMemcpyDeviceToHost);
+	cudaFree(d_result);
+	tmp.clear();
+}
 
 GKO_INSTANTIATE_FOR_EACH_VALUE_TYPE(GKO_DECLARE_DENSE_COUNT_NONZEROS_KERNEL);
 
+namespace kernel
+{
+
+__global__ __launch_bounds__(default_block_size) void get_max(size_type num_rows, size_type* nnz_per_row,
+		size_type* result, size_type start_step)
+{
+	const auto tidx = threadIdx.x + blockIdx.x * blockDim.x;
+	for (auto step = start_step; step >= 1; step/=2) {
+		if (tidx < step && tidx + step < num_rows) {
+			nnz_per_row[tidx] = max(nnz_per_row[tidx], nnz_per_row[tidx + step]);
+		}
+		__syncthreads();
```

Comment:
You can only synchronize inside a block, never over all blocks.
I would recommend doing this in two steps, one that uses all blocks and reduces it to an amount that will be processed in the next kernel launch with only a single block.